### PR TITLE
chore: migrate Tooltip

### PR DESCRIPTION
**Pull request by RobelTekle**
## DESCRIPTION
Migrate Tooltip, same api as before.

## SCREENSHOTS / SCREEN RECORDINGS

https://github.com/user-attachments/assets/78333f28-0c21-4f11-a83a-f9aeb9990ba6

## COMPATIBILITY

- [ ] Tested on Safari (desktop)
- [ ] Tested on Chrome (desktop)
- [ ] Tested on Firefox (desktop)
- [ ] Tested on mobile device sizes
- [ ] Tested on tablet device sizes
- [ ] Tested on IOS Safari (either device or simulator)

## QA

- [ ] Thoroughly tested in local environment
- [ ] Added tests for all new features
- [ ] Added tests that considered edge cases

### CODE REVIEW

File: lib/src/components/Tooltip/docs/examples/placement.tsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import * as React from 'react'
+
+import { Button } from '@/components/Button'
+import { Tooltip } from '@/components/Tooltip'
+import { Checkbox } from '@old/Checkbox'
+import { Field } from '@old/Field'
+import { Flex } from '@old/Flex'
+import { InputText } from '@old/InputText'
```
Comment (cnairi) [inline lib/src/components/Tooltip/docs/examples/placement.tsx]: For InputText, we already have it in the new version

File: lib/src/components/Tooltip/docs/examples/placement.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import * as React from 'react'
+
+import { Button } from '@/components/Button'
+import { Tooltip } from '@/components/Tooltip'
+import { Checkbox } from '@old/Checkbox'
+import { Field } from '@old/Field'
+import { Flex } from '@old/Flex'
```
Comment (cnairi) [inline lib/src/components/Tooltip/docs/examples/placement.tsx]: You can safely remove Flex as we decided to depreacte ite

File: lib/src/components/Tooltip/docs/examples/placement.tsx [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import { InputText } from '@old/InputText'
+const Example = () => {
+  const [withArrow, setWithArrow] = React.useState(true)
+  const [tooltipContent, setTooltipContent] = React.useState('This is the tooltip content')
+
+  return (
+    <Flex flexDirection="column" gap="lg">
```
Comment (cnairi) [inline lib/src/components/Tooltip/docs/examples/placement.tsx]: `<div className="flex flex-col gap-lg">`

File: lib/src/components/Tooltip/docs/examples/placement.tsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      </Field>
+
+      <Field label="Tooltip content">
+        <InputText
+          name="tooltipContent"
+          onChange={e => setTooltipContent(e.target.value)}
+          value={tooltipContent}
+        />
+      </Field>
+      <Flex flexWrap="wrap" gap="lg">
```
Comment (cnairi) [inline lib/src/components/Tooltip/docs/examples/placement.tsx]: `<div className="flex flex-wrap gap-lg"`

File: lib/src/components/Tooltip/docs/examples/placement.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            <span>top &#8593; </span>
+          </Button>
+        </Tooltip>
+        <Tooltip content={tooltipContent} fixed placement="top-end" withArrow={withArrow}>
+          <Button className="w-[120px]" size="sm">
+            <span>top-end &#8599; </span>
+          </Button>
+        </Tooltip>
+      </Flex>
+      <Flex flexWrap="wrap" gap="lg">
```
Comment (cnairi) [inline lib/src/components/Tooltip/docs/examples/placement.tsx]: `<div className="flex flex-wrap gap-lg"`

File: lib/src/components/Tooltip/docs/examples/placement.tsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            <span>bottom &#8595; </span>
+          </Button>
+        </Tooltip>
+        <Tooltip content={tooltipContent} fixed placement="bottom-end" withArrow={withArrow}>
+          <Button className="w-[120px]" size="sm">
+            <span>bottom-end &#8600; </span>
+          </Button>
+        </Tooltip>
+      </Flex>
+      <Flex flexWrap="wrap" gap="lg">
```
Comment (cnairi) [inline lib/src/components/Tooltip/docs/examples/placement.tsx]: `<div className="flex flex-wrap gap-lg"`

File: lib/src/components/Tooltip/docs/examples/placement.tsx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            <span>left &#8592; </span>
+          </Button>
+        </Tooltip>
+        <Tooltip content={tooltipContent} fixed placement="left-end" withArrow={withArrow}>
+          <Button className="w-[120px]" size="sm">
+            <span>left-end &#8601; </span>
+          </Button>
+        </Tooltip>
+      </Flex>
+      <Flex flexWrap="wrap" gap="lg">
```
Comment (cnairi) [inline lib/src/components/Tooltip/docs/examples/placement.tsx]: `<div className="flex flex-wrap gap-lg"`

File: lib/src/components/Tooltip/tooltip.module.scss [hunk 8]
```diff
@@ -0,0 +1,166 @@
+@layer components {
```
Comment (cnairi) [inline lib/src/components/Tooltip/tooltip.module.scss:1]: I think you can factorize even more the css variables, to identify exactly which one will be modified by the different variants:
```
@layer components {
  .root {
    background-color: var(--color-neutral-90);
    border: var(--border-width-sm) solid var(--color-neutral-60);
    border-radius: var(--radius-md);
    box-sizing: border-box;
    color: var(--color-neutral-10);
    font-size: var(--font-size-xs);
    font-weight: var(--font-weight-medium);
    max-width: 12.5rem;
    padding: var(--spacing-xs) var(--spacing-sm);
    transition-property: opacity, transform, visibility;
    transition-duration: var(--duration-medium);
    transform-origin: top center;
    z-index: 10;
    visibility: var(--visibility, hidden);
    opacity: var(--opacity, 0);
    transform: var(--transform, none);

    
    [data-enter] & {
      --transform: translate3d(0, 0, 0);
      --visibility: visible;
      --opacity: 1;
    }
  }

  /* With arrow variant */
  .withArrow {
    &::after {
      content: '';
      position: absolute;
      background-color: inherit;
      border-width: inherit;
      border-style: solid;
      border-color: transparent var(--color-neutral-60) var(--color-neutral-60) transparent;
      border-bottom-right-radius: inherit;
      clip-path: polygon(0% 100%, 0% 100%, 100% 100%, 100% 0%);
      z-index: 0;
      width: var(--tooltipArrowSize);
      height: var(--tooltipArrowSize);
      translate: var(--withArrowTranslate, -50% -50%);
      rotate: var(--withArrowRotate, 0);
      left: var(--withArrowLeft, auto);
      top: var(--withArrowTop, auto);
    }
  }

  /* Placement variants with arrows */
  .placement {
    /* Top placements */
    &-top,
    &-top-start,
    &-top-end {
      &.withArrow::after {
        --withArrowTop: 100%;
        --withArrowRotate: 45deg;
      }
    }

    &-top.withArrow::after {
      --withArrowLeft: 50%
    }

    &-top-start.withArrow::after {
      --withArrowLeft: var(--tooltipArrowSpacer);
    }

    &-top-end.withArrow::after {
      --withArrowLeft: calc(100% - var(--tooltipArrowSpacer));
    }

    /* Bottom placements */
    &-bottom,
    &-bottom-start,
    &-bottom-end {
      &.withArrow::after {
        --withArrowTranslate: -50% 50%;
        --withArrowRotate: 225deg;
        bottom: 100%;
      }
    }

    &-bottom.withArrow::after {
      --withArrowLeft: 50%;
    }

    &-bottom-start.withArrow::after {
      --withArrowLeft: var(--tooltipArrowSpacer);
    }

    &-bottom-end.withArrow::after {
      --withArrowLeft: calc(100% - var(--tooltipArrowSpacer));
    }

    /* Left placements */
    &-left,
    &-left-start,
    &-left-end {
      &.withArrow::after {
        --withArrowLeft: 100%;
        --withArrowTop: var(--tooltipYPosition);
        --withArrowRotate: -45deg;
      }
    }

    /* Right placements */
    &-right,
    &-right-start,
    &-right-end {
      &.withArrow::after {
        --withArrowLeft: 0;
        --withArrowTop: var(--tooltipYPosition);
        --withArrowRotate: 135deg;
      }
    }
  }

  /* Transform directions for fixed tooltips */
  .transform {
    &-auto {
      --transform: translate3d(0, 0, 0);
    }

    &-top {
      --transform: translate3d(0, -4px, 0);
    }

    &-bottom {
      --transform: translate3d(0, 4px, 0);
    }

    &-left {
      --transform: translate3d(-4px, 0, 0);
    }

    &-right {
      --transform: translate3d(4px, 0, 0);
    }

    &-top-start,
    &-top-end {
      --transform: translate3d(-4px, -4px, 0);
    }

    &-bottom-start,
    &-bottom-end {
      --transform: translate3d(-4px, 4px, 0);
    }

    &-left-start,
    &-left-end {
      --transform: translate3d(-4px, -4px, 0);
    }

    &-right-start,
    &-right-end {
      --transform: translate3d(4px, -4px, 0);
    }

    &-right-end,
    &-bottom-end {
      --transform: translate3d(4px, 4px, 0);
    }
  }

  .childItem {
    display: inline-block;
  }
}
```

File: lib/src/components/Tooltip/index.tsx [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+              {
+                '--tooltipArrowSize': `${ARROW_SIZE}px`,
+                '--tooltipArrowSpacer': `${ARROW_SPACER}px`,
+                '--tooltipYPosition': getYPosition(currentPlacement, popoverHeight),
+              } as React.CSSProperties & Record<string, number | string>
+            }
+          >
+            {content}
+          </div>
+        </AriakitTooltip>
```
Comment (cnairi) [inline lib/src/components/Tooltip/index.tsx:107]: I just had a look an I really think the only custom property we need is on:
- Updating the placement when we want our tooltip to follow our mouse
- Animation that is a bit harsh on ariakit

```
import {
  Tooltip as AriakitTooltip,
  TooltipAnchor as AriakitTooltipAnchor,
  TooltipArrow,
  useStoreState,
  useTooltipStore,
} from '@ariakit/react'
import type {
  TooltipProps as AriakitTooltipProps,
  TooltipStoreProps as AriakitTooltipStoreProps,
} from '@ariakit/react'
import React, { forwardRef, useCallback, useEffect, useState } from 'react'

import { classNames } from '@/utils'

import tooltipStyles from './tooltip.module.scss'

export interface TooltipOptions {
  children: React.ReactNode
  content: JSX.Element | string
  fixed?: boolean
  gutter?: number
  placement?: TooltipPlacement
  withArrow?: boolean
}

export type TooltipPlacement = AriakitTooltipStoreProps['placement']

export type TooltipProps = Omit<AriakitTooltipProps, 'children' | 'content'> & TooltipOptions

const cx = classNames(tooltipStyles)

export const Tooltip = forwardRef<HTMLDivElement, TooltipProps>(
  (
    {
      children,
      className = '',
      content,
      fixed = false,
      gutter = 8,
      placement = fixed ? 'top' : 'bottom',
      withArrow = false,
      ...rest
    }: TooltipProps,
    ref
  ) => {
    const tooltip = useTooltipStore({ animated: 300, placement })
    const { anchorElement, mounted, popoverElement } = useStoreState(tooltip)
    const { render, setState } = tooltip
    const [position, setPosition] = useState({ x: 0, y: 0 })

    const updatePosition = () => {
      if (!popoverElement) return

      Object.assign(popoverElement.style, {
        display: mounted ? 'block' : 'none',
        left: `${position.x}px`,
        position: 'absolute',
        top: `${position.y + window.scrollY + 20}px`,
      })
    }

    const stopAnimation = useCallback(() => {
      setState('animating', false)
    }, [setState])

    // Handle disabled children by wrapping them
    const child = (children as JSX.Element)?.props?.disabled
      ? React.Children.only(<span className={cx('childItem')}>{children}</span>)
      : children

    useEffect(() => {
      function onMouseMove({ clientX, clientY }: { clientX: number; clientY: number }) {
        setPosition({ x: clientX, y: clientY })
        render()
      }

      if (anchorElement && !fixed) {
        anchorElement.addEventListener('mousemove', onMouseMove)

        return () => {
          anchorElement.removeEventListener('mousemove', onMouseMove)
        }
      }
    }, [render, fixed, anchorElement])

    // If no content, simply return the children
    if (!content) {
      return children as React.ReactElement
    }

    return (
      <>
        <AriakitTooltipAnchor render={child as React.ReactElement} store={tooltip} />
        <AriakitTooltip
          className={cx('root', className)}
          fixed={fixed}
          gutter={gutter}
          onTransitionEnd={stopAnimation}
          ref={ref}
          render={<div {...rest} />}
          store={tooltip}
          updatePosition={fixed ? undefined : updatePosition}
        >
          {withArrow ? <TooltipArrow className={cx('withArrow')} /> : null}
          {content}
        </AriakitTooltip>
      </>
    )
  }
)
```

As a result, the utils file is no longer useful.

And your scss file is even simpler:
```
@layer components {
  .root {
    background-color: var(--color-neutral-90);
    border: var(--border-width-sm) solid var(--color-neutral-60);
    border-radius: var(--radius-md);
    box-sizing: border-box;
    color: var(--color-neutral-10);
    font-size: var(--font-size-xs);
    font-weight: var(--font-weight-medium);
    max-width: 12.5rem;
    padding: var(--spacing-xs) var(--spacing-sm);
    transition-property: opacity, transform, visibility;
    transition-duration: var(--duration-medium);
    transform-origin: top center;
    z-index: 10;
    visibility: var(--visibility, hidden);
    opacity: var(--opacity, 0);
    transform: var(--transform, none);


    &[data-enter="true"] {
      --transform: translate3d(0, 0, 0);
      --visibility: visible;
      --opacity: 1;
    }
  }

  /* With arrow variant */
  .withArrow {
    &::after {
      content: '';
      position: absolute;
      background-color: inherit;
      border-width: inherit;
      border-style: solid;
      border-color: transparent var(--color-neutral-60) var(--color-neutral-60) transparent;
      border-bottom-right-radius: inherit;
      clip-path: polygon(0% 100%, 0% 100%, 100% 100%, 100% 0%);
      z-index: 0;
      width: 12px;
      height: 12px;
      translate: var(--withArrowTranslate, -50% -50%);
      rotate: var(--withArrowRotate, 0);
      left: var(--withArrowLeft, auto);
      top: var(--withArrowTop, auto);
    }
  }

  

  .childItem {
    display: inline-block;
  }
}
```

You can decide if you want to implement it now or later during the redesign phase if it seems to much in your opinion

File: lib/src/components/Tooltip/index.tsx [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+          gutter={gutter}
+          ref={ref}
+          render={<div {...rest} />}
+          store={tooltip}
+          updatePosition={fixed ? undefined : updatePosition}
+        >
+          <div
+            className={cx(
+              'root',
+              withArrow && 'withArrow',
```
Comment (anucreative) [inline lib/src/components/Tooltip/index.tsx]: Can we use https://ariakit.org/reference/tooltip-arrow so we don't have to worry about styling / positioning it ourselves ? 

https://ariakit.org/reference/tooltip-arrow

Would mean we could get rid of lots of styles and `getYPosition` etc.

**Review by anucreative — APPROVED** [review]
Just a small question about https://ariakit.org/reference/tooltip-arrow

File: lib/src/components/Tooltip/index.tsx (hunk 9 above)
Comment (RobelTekle) [reply lib/src/components/Tooltip/index.tsx:107]: You are totally right

**Comment by RobelTekle** [comment]
@anucreative @cnairi I get rid off all arrow logic and use Ariakit tooltip arrow component. Thanks for your feedback 🤗
https://github.com/WTTJ/welcome-ui/pull/2853/commits/732000299599f9f354985e256156507e73b38fdf
